File: packages/core/src/export/doacao.ts
import { eq } from "drizzle-orm";

import {
  type ContaBancaria,
  type Db,
  type Movimentacao,
  type MovimentacaoSpca,
  doacaoFinanceiraLink,
} from "@spc-up/db";

import {
  buildCabecalho,
  exportPath,
  formatMoeda,
  makeDoacaoRoot,
  sub,
  writeXml,
  xmlToBuffer,
} from "./common";

const DEFAULT_NUM_PARTIDO = 13;
const FORMA_DOACAO: Record<string, string> = {
  PIX: "PIX",
  TED: "TED",
  TEL: "TEL",
  TEB: "TEB",
  CH: "CH",
};
const OPERACAO_FINANCEIRA: Record<string, string> = {
  PIX: "PIX",
  TED: "TED",
  TEL: "TEL",
  EP: "EP",
};

export type MovimentacaoDoacao = Movimentacao & {
  spca: MovimentacaoSpca;
  pessoaFisica: { cpf: string; nome: string; tituloEleitor: string | null } | null;
  contaBancaria: ContaBancaria | null;
};

async function fetchMovimentacoes(
  db: Db,
  uf: string,
  exercicio: number,
): Promise<MovimentacaoDoacao[]> {
  const ufUpper = uf.toUpperCase();
  const links = await db.query.doacaoFinanceiraLink.findMany({
    where: eq(doacaoFinanceiraLink.sincronizado, true),
    with: {
      movimentacaoOrigem: {
        with: { spca: true, pessoaFisica: true, contaBancaria: true },
      },
    },
  });

  return links
    .map((link) => link.movimentacaoOrigem)
    .filter(
      (mov) =>
        mov != null &&
        mov.uf === ufUpper &&
        mov.exercicio === exercicio &&
        mov.direcao === "ENTRADA" &&
        mov.status === "CONFIRMADO" &&
        mov.spca != null,
    )
    .sort((a, b) => {
      const byDate = String(a.dataMovimento).localeCompare(String(b.dataMovimento));
      return byDate !== 0 ? byDate : a.id.localeCompare(b.id);
    }) as MovimentacaoDoacao[];
}

function formatDate(value: string | Date): string {
  if (typeof value === "string") {
    return value.slice(0, 10);
  }
  return value.toISOString().slice(0, 10);
}

function appendContaDestino(
  parent: ReturnType<typeof sub>,
  conta: ContaBancaria | null,
): void {
  const contaDestino = sub(parent, "contaBancariaDestino");
  const banco = sub(contaDestino, "bancoDestino");
  if (conta == null) {
    sub(banco, "nrBancoOrigem", "000");
    sub(banco, "agenciaOrigem", "0000");
    sub(banco, "contaCorrente", "0");
    sub(banco, "dvContaCorrente", "0");
    return;
  }
  sub(banco, "nrBancoOrigem", "000");
  sub(banco, "agenciaOrigem", conta.agencia);
  sub(banco, "contaCorrente", conta.conta);
  sub(banco, "dvContaCorrente", conta.dv ?? "0");
}

function appendBeneficiario(
  parent: ReturnType<typeof sub>,
  options: { cnpj: string; uf: string; exercicio: number },
): void {
  const beneficiario = sub(parent, "beneficiario");
  const partido = sub(beneficiario, "partido");
  sub(partido, "tipo", "PP");
  sub(partido, "nrCnpj", options.cnpj);
  sub(partido, "esferaPartidaria", "ESTADUAL");
  sub(partido, "eleicao", options.exercicio);
  sub(partido, "eleicaoSuplementar", "N");
  sub(partido, "partido", DEFAULT_NUM_PARTIDO);
  sub(partido, "uf", options.uf);
}

function appendDoador(
  parent: ReturnType<typeof sub>,
  mov: MovimentacaoDoacao,
  spca: MovimentacaoSpca,
): void {
  const pessoa = mov.pessoaFisica;
  if (pessoa == null) {
    throw new Error(`Movimentacao ${mov.id} requires pessoa_fisica for doacao export`);
  }
  const recibo = spca.nrReciboDoacao;
  if (!recibo) {
    throw new Error(`Movimentacao ${mov.id} missing nr_recibo_doacao`);
  }

  const doadores = sub(parent, "doadoresOriginarios");
  const dador = sub(doadores, "dadorOriginario");
  const pf = sub(dador, "pessoaFisica");
  sub(pf, "nrCpf", pessoa.cpf);
  sub(pf, "nmPessoa", pessoa.nome);
  if (pessoa.tituloEleitor) {
    sub(pf, "tituloEleitor", pessoa.tituloEleitor);
  }
  sub(pf, "nrReciboDoacao", recibo);
  sub(pf, "vrDoacao", formatMoeda(mov.valor));
}

function appendDoacaoItem(
  parent: ReturnType<typeof sub>,
  mov: MovimentacaoDoacao,
  options: { cnpj: string; uf: string; exercicio: number },
): void {
  const spca = mov.spca;
  const doacao = sub(parent, "doacao");
  appendBeneficiario(doacao, options);
  sub(doacao, "dtDoacao", formatDate(mov.dataMovimento));
  sub(doacao, "fonteRecurso", spca.fonteRecurso ?? "OR");

  const classificacao = sub(doacao, "classificacoesDoacao");
  sub(classificacao, "valorDoacao", formatMoeda(mov.valor));

  const especie = (spca.especieRecurso ?? "PIX").toUpperCase();
  sub(doacao, "formaDoacao", FORMA_DOACAO[especie] ?? "PIX");
  const operacao = OPERACAO_FINANCEIRA[especie];
  if (operacao) {
    sub(doacao, "operacaoFinanceira", operacao);
  }

  const nrExtrato = mov.nrExtratoBancario ?? spca.nrReciboDoacao ?? "0";
  sub(doacao, "nrExtratoBancario", nrExtrato);
  sub(doacao, "contaBancariaOrigem");
  sub(doacao, "nrDocumento", nrExtrato);
  sub(doacao, "nrReciboDoacao", spca.nrReciboDoacao);
  appendDoador(doacao, mov, spca);
  appendContaDestino(doacao, mov.contaBancaria);
}

export function buildDoacaoDocument(
  movimentacoes: MovimentacaoDoacao[],
  cnpj: string,
  uf: string,
  exercicio: number,
) {
  const root = makeDoacaoRoot();
  buildCabecalho(root, { cnpj, exercicio });
  const corpo = sub(root, "CORPO");
  const doacoes = sub(corpo, "doacoes");
  sub(doacoes, "totalDoacao", movimentacoes.length);
  const ufUpper = uf.toUpperCase();
  for (const mov of movimentacoes) {
    appendDoacaoItem(doacoes, mov, { cnpj, uf: ufUpper, exercicio });
  }
  return root;
}

export async function buildDoacaoXml(
  db: Db,
  uf: string,
  exercicio: number,
  cnpj: string,
): Promise<string> {
  const movimentacoes = await fetchMovimentacoes(db, uf, exercicio);
  const doc = buildDoacaoDocument(movimentacoes, cnpj, uf, exercicio);
  return writeXml(doc, exportPath(uf.toUpperCase(), exercicio, cnpj, "doacao"));
}

export async function buildDoacaoXmlBuffer(
  db: Db,
  uf: string,
  exercicio: number,
  cnpj: string,
): Promise<Buffer> {
  const movimentacoes = await fetchMovimentacoes(db, uf, exercicio);
  return xmlToBuffer(buildDoacaoDocument(movimentacoes, cnpj, uf, exercicio));
}

File: legacy/python/spc_up/services/export/doacao.py
"""Build Doacao Financeira XML from synchronized doacao links."""

from __future__ import annotations

from pathlib import Path

from lxml import etree
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from spc_up.config import settings
from spc_up.models.entities import (
    ContaBancaria,
    DoacaoFinanceiraLink,
    Movimentacao,
    MovimentacaoDirecao,
    MovimentacaoSpca,
    MovimentacaoStatus,
    PessoaFisica,
)
from spc_up.services.export.common import format_moeda, write_xml

DOACAO_NS = "http://www.tse.jus.br/2012/XMLSchema/doacaoFinanceiraPartidoCandidato.xsd"
DOACAO_NSMAP = {None: DOACAO_NS}

_DEFAULT_NUM_PARTIDO = 13
_FORMA_DOACAO = {"PIX": "PIX", "TED": "TED", "TEL": "TEL", "TEB": "TEB", "CH": "CH"}
_OPERACAO_FINANCEIRA = {"PIX": "PIX", "TED": "TED", "TEL": "TEL", "EP": "EP"}


def _sub(parent: etree._Element, tag: str, text: str | int | None = None) -> etree._Element:
    element = etree.SubElement(parent, f"{{{DOACAO_NS}}}{tag}")
    if text is not None:
        element.text = str(text)
    return element


def make_doacao_root() -> etree._Element:
    return etree.Element(f"{{{DOACAO_NS}}}spcaImportacaoArquivo", nsmap=DOACAO_NSMAP)


def _build_cabecalho(parent: etree._Element, *, cnpj: str, exercicio: int) -> None:
    cabecalho = _sub(parent, "CABECALHO")
    _sub(cabecalho, "nrCnpjPrestador", cnpj)
    _sub(cabecalho, "anoExercicio", exercicio)


def _fetch_movimentacoes(session: Session, uf: str, exercicio: int) -> list[Movimentacao]:
    return list(
        session.execute(
            select(Movimentacao)
            .join(DoacaoFinanceiraLink, DoacaoFinanceiraLink.movimentacao_origem_id == Movimentacao.id)
            .join(MovimentacaoSpca, MovimentacaoSpca.movimentacao_id == Movimentacao.id)
            .options(
                selectinload(Movimentacao.spca),
                selectinload(Movimentacao.pessoa_fisica),
                selectinload(Movimentacao.conta_bancaria),
                selectinload(Movimentacao.doacao_link),
            )
            .where(
                Movimentacao.uf == uf,
                Movimentacao.exercicio == exercicio,
                Movimentacao.direcao == MovimentacaoDirecao.ENTRADA.value,
                Movimentacao.status == MovimentacaoStatus.CONFIRMADO.value,
                DoacaoFinanceiraLink.sincronizado.is_(True),
            )
            .order_by(Movimentacao.data_movimento, Movimentacao.id)
        ).scalars().all()
    )


def _append_conta_destino(parent: etree._Element, conta: ContaBancaria | None) -> None:
    conta_destino = _sub(parent, "contaBancariaDestino")
    banco = _sub(conta_destino, "bancoDestino")
    if conta is None:
        _sub(banco, "nrBancoOrigem", "000")
        _sub(banco, "agenciaOrigem", "0000")
        _sub(banco, "contaCorrente", "0")
        _sub(banco, "dvContaCorrente", "0")
        return

    _sub(banco, "nrBancoOrigem", "000")
    _sub(banco, "agenciaOrigem", conta.agencia)
    _sub(banco, "contaCorrente", conta.conta)
    _sub(banco, "dvContaCorrente", conta.dv or "0")


def _append_beneficiario(parent: etree._Element, *, cnpj: str, uf: str, exercicio: int) -> None:
    beneficiario = _sub(parent, "beneficiario")
    partido = _sub(beneficiario, "partido")
    _sub(partido, "tipo", "PP")
    _sub(partido, "nrCnpj", cnpj)
    _sub(partido, "esferaPartidaria", "ESTADUAL")
    _sub(partido, "eleicao", exercicio)
    _sub(partido, "eleicaoSuplementar", "N")
    _sub(partido, "partido", _DEFAULT_NUM_PARTIDO)
    _sub(partido, "uf", uf)


def _append_doador(parent: etree._Element, movimentacao: Movimentacao, spca: MovimentacaoSpca) -> None:
    pessoa: PessoaFisica | None = movimentacao.pessoa_fisica
    if pessoa is None:
        raise ValueError(f"Movimentacao {movimentacao.id} requires pessoa_fisica for doacao export")

    recibo = spca.nr_recibo_doacao
    if not recibo:
        raise ValueError(f"Movimentacao {movimentacao.id} missing nr_recibo_doacao")

    doadores = _sub(parent, "doadoresOriginarios")
    dador = _sub(doadores, "dadorOriginario")
    pf = _sub(dador, "pessoaFisica")
    _sub(pf, "nrCpf", pessoa.cpf)
    _sub(pf, "nmPessoa", pessoa.nome)
    if pessoa.titulo_eleitor:
        _sub(pf, "tituloEleitor", pessoa.titulo_eleitor)
    _sub(pf, "nrReciboDoacao", recibo)
    _sub(pf, "vrDoacao", format_moeda(movimentacao.valor))


def _append_doacao_item(
    parent: etree._Element,
    movimentacao: Movimentacao,
    *,
    cnpj: str,
    uf: str,
    exercicio: int,
) -> None:
    spca = movimentacao.spca
    if spca is None:
        raise ValueError(f"Movimentacao {movimentacao.id} missing spca payload")

    doacao = _sub(parent, "doacao")
    _append_beneficiario(doacao, cnpj=cnpj, uf=uf, exercicio=exercicio)
    _sub(doacao, "dtDoacao", movimentacao.data_movimento.isoformat())
    _sub(doacao, "fonteRecurso", spca.fonte_recurso or "OR")

    classificacao = _sub(doacao, "classificacoesDoacao")
    _sub(classificacao, "valorDoacao", format_moeda(movimentacao.valor))

    especie = (spca.especie_recurso or "PIX").upper()
    forma = _FORMA_DOACAO.get(especie, "PIX")
    _sub(doacao, "formaDoacao", forma)
    operacao = _OPERACAO_FINANCEIRA.get(especie)
    if operacao:
        _sub(doacao, "operacaoFinanceira", operacao)

    nr_extrato = movimentacao.nr_extrato_bancario or spca.nr_recibo_doacao or "0"
    _sub(doacao, "nrExtratoBancario", nr_extrato)
    _sub(doacao, "contaBancariaOrigem")
    _sub(doacao, "nrDocumento", nr_extrato)
    _sub(doacao, "nrReciboDoacao", spca.nr_recibo_doacao)
    _append_doador(doacao, movimentacao, spca)
    _append_conta_destino(doacao, movimentacao.conta_bancaria)


def build_doacao_xml(session: Session, uf: str, exercicio: int, cnpj: str) -> Path:
    """Build and persist Doacao Financeira XML for synchronized doacao links."""
    movimentacoes = _fetch_movimentacoes(session, uf, exercicio)

    root = make_doacao_root()
    _build_cabecalho(root, cnpj=cnpj, exercicio=exercicio)

    corpo = _sub(root, "CORPO")
    doacoes = _sub(corpo, "doacoes")
    _sub(doacoes, "totalDoacao", len(movimentacoes))

    for movimentacao in movimentacoes:
        _append_doacao_item(doacoes, movimentacao, cnpj=cnpj, uf=uf, exercicio=exercicio)

    output_path = (
        Path(settings.storage_root)
        / "exports"
        / uf
        / str(exercicio)
        / f"doacao_{cnpj}.xml"
    )
    return write_xml(root, output_path)
